### ISSUE TITLE
Odd issue with evaluating LongPtr as Boolean

### ISSUE THREAD

**Issue by fafalone** [twinbasic/twinbasic#1143]
**Describe the bug**
I have a macro for the Header Control in a ListView; where SendMessage returns a LongPtr:
```
Public Function Header_GetItem(hwndHD As LongPtr, iItem As Long, phdi As HDITEM) As Boolean
  Header_GetItem = SendMessage(hwndHD, HDM_GETITEM, iItem, phdi)
End Function
```
In VB6, this is allowed, and works as normal.

In tB, this sometimes raises an error (it was for quite a while but it stopped and now I can't reproduce it; it was some crazy application-defined error), and the call always fails; I'm using it to request an lParam, which always comes back as zero.

**To Reproduce**
I'm attaching a debug build of my TBEventTrace project, since building an API created control to SendMessage to is a pain. It's all set to go; one needs only to run the project. It will automatically generate test data to display in the ListView, you do not need to start a trace or click anything. You will observe every single item and column shows the same text; this is due to the LVN_GETDISPINFO handler in modSubclass' call to modListViewAndHeader.GetHDItemlParam failing, which you'll see in the Debug.Print output, if it doesn't show the error I was getting (they'll all be 0, they should be nonzero).

GetHDItemlParam calls the macro, and there the problem is explained:
```
Public Function Header_GetItem(hwndHD As LongPtr, iItem As Long, phdi As HDITEM) As Boolean
  'This fails:
  Header_GetItem = SendMessage(hwndHD, HDM_GETITEM, iItem, phdi)
  'This works:
  'Header_GetItem = (SendMessage(hwndHD, HDM_GETITEM, iItem, phdi) <> 0)
End Function
```
If you switch which version is used by commenting out the first and uncommenting the second call, you'll see the expected different text in different columns/items and non-zero returns in the debug console.

**Expected behavior**
This type of casting should be allowed as it is in VB. 

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 10
 - twinBASIC compiler version 0.15.108

**Additional context**
The full sequence is:
1. The ListView is created and subclassed in Form1.InitLV after a call in Form_Load; the WndProc is F1WndProc in modSubclass, which forwards WM_NOTIFY to DoLVNotify in same module. 
2. Since the ListView has LVS_OWNERDATA set, it generates LVN_GETDISPINFOW messages to request the text.
3. I added new optional columns, so started storing the column id in the lParam, so it calls GetHDItemlParam in modListViewAndHeader.
4. That function calls the Header_GetItem macro where the problem is.

[TBEventTrace-Dbg.zip](https://github.com/twinbasic/twinbasic/files/9484914/TBEventTrace-Dbg.zip)

**Comment by fafalone** [comment]
The application defined error is back! And this time it's impacting the version that was working too. I didn't change anything besides turning off the Debug.Print statement while working on isolating another bug. 

Run-time error -1379947480 (ADBFAC28) Application-defined or object-defined error.

What I did was switch to 32bit, run it, switch to 64bit, delete the debug.print statement in GetHDItemlParam, then run it (in the IDE, all times here). I can't guarantee that will reproduce it, but such is the nature of bizarre bugs. But that it's doing this despite the previously-working alternative means I no longer have a workaround and my project is broken until this error stops again. 

I put the Debug.Print statement back in, same line different error:

Run-time error -1379947560 (ADBFABD8)  Application-defined or object-defined error.

I took it back out (this time by commenting it out), it returned to the other error. Put it back in again, it went back to the 2nd error.

**Comment by WaynePhillipsEA** [comment]
Is this affecting both 32-bit and 64-bit?

**Comment by fafalone** [comment]
Yes, although the runtime error went away again when I switched back to 32bit. 

Then when I switched back to 64bit, the runtime error was back with a whole new number.

Run-time error -2107661224 (825FA458)

**Comment by WaynePhillipsEA** [comment]
I think the strange runtime errors here might be due to the incorrect usage of `StrPtr()` as I just mentioned over on discord.

For the main issue, the problem is not technically the return value.  The problem here is that the ByRef UDT is not being populated.  This is due to the implicit ANSI DLL conversions introducing a temporary ANSI version of the UDT when SendMessage is called (for the `phdi` argument) combined with a return-value optimization that the compiler implements when  you use the new `Return` syntax.  Even though you didn't explicitly use the `Return` syntax, the compiler chose to optimize the tail of your function into a `Return` statement, because it was a simple assignment to the function return variable, followed only by an Exit/End Sub.  The optimized `Return` syntax is currently preventing the conversion of the temporary ANSI-UDT back to the real UDT, and so the real UDT remains unmodified.

In short, this will be fixed in the next release ;)

**Comment by WaynePhillipsEA** [comment]
This is fixed in BETA 109, thanks!
